File: src/main/java/me/testaccount666/serversystem/userdata/UserManager.java
package me.testaccount666.serversystem.userdata;

import me.testaccount666.serversystem.ServerSystem;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.io.File;
import java.nio.file.Path;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Manages ServerSystem's user data.
 * This class is responsible for creating, caching, and retrieving user objects,
 * as well as cleaning up stale user data to prevent memory leaks.
 */
public class UserManager {
    public static final Path USER_DATA_PATH = Path.of("plugins", "ServerSystem", "UserData");
    private static final ConsoleUser _CONSOLE_USER = new ConsoleUser();
    private final Map<String, CachedUser> _userMap = new ConcurrentHashMap<>();
    private final Map<UUID, CachedUser> _userUuidMap = new ConcurrentHashMap<>();

    public UserManager() {
        Bukkit.getScheduler().scheduleAsyncRepeatingTask(ServerSystem.Instance, this::cleanStaleUsers, 15 * 20 * 60, 15 * 20 * 60); // 15 Minutes

        Bukkit.getOnlinePlayers().forEach(this::getUser);
    }

    private static File getUserFile(UUID uuid) {
        return USER_DATA_PATH.resolve("${uuid}.yml.gz").toFile();
    }

    /**
     * Gets the console user instance.
     *
     * @return The console user instance
     */
    public static ConsoleUser getConsoleUser() {
        return _CONSOLE_USER;
    }

    /**
     * Gets a cached user by player instance.
     *
     * @param player The player to get the user for
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    public Optional<CachedUser> getUser(Player player) {
        if (player.hasMetadata("NPC")) return Optional.of(new CachedUser(new NpcUser()));

        return getUser(player.getUniqueId(), true);
    }

    /**
     * Gets a cached user by UUID.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param uuid The UUID of the user to get
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    public Optional<CachedUser> getUser(UUID uuid) {
        return getUser(uuid, false);
    }

    /**
     * Gets a cached user by UUID.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param uuid            The UUID of the user to get
     * @param forceOnlineUser If the method should only return online users
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    public Optional<CachedUser> getUser(UUID uuid, boolean forceOnlineUser) {
        if (!forceOnlineUser && uuid.equals(ConsoleUser.CONSOLE_UUID)) return Optional.of(new CachedUser(getConsoleUser()));

        if (_userUuidMap.containsKey(uuid)) {
            var cachedUser = _userUuidMap.get(uuid);
            if (cachedUser.isOfflineUser() && forceOnlineUser) return Optional.empty();

            cachedUser.updateLastAccessTime();

            return Optional.of(cachedUser);
        }

        var player = Bukkit.getPlayer(uuid);
        if (player != null) return Optional.of(createOnlineUser(uuid));
        if (forceOnlineUser) return Optional.empty();

        return createOfflineUser(uuid);
    }

    /**
     * Gets a cached user by name.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param name The name of the user to get
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    public Optional<CachedUser> getUser(String name) {
        return getUser(name, false);
    }

    /**
     * Gets a cached user by name.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param name            The name of the user to get
     * @param forceOnlineUser If the method should only return online users
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    public Optional<CachedUser> getUser(String name, boolean forceOnlineUser) {
        if (_userMap.containsKey(name)) {
            var cachedUser = _userMap.get(name);

            if (cachedUser.isOfflineUser() && forceOnlineUser) return Optional.empty();

            cachedUser.updateLastAccessTime();

            return Optional.of(cachedUser);
        }

        var player = Bukkit.getPlayer(name);
        if (player != null) return Optional.of(createOnlineUser(player.getUniqueId()));
        if (forceOnlineUser) return Optional.empty();

        var offlineUser = Bukkit.getOfflinePlayer(name);
        if (offlineUser.getName() == null) return Optional.empty();

        return createOfflineUser(offlineUser.getUniqueId());
    }

    /**
     * Creates and caches an offline user for the given UUID.
     *
     * @param uuid The UUID of the user to create
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    private Optional<CachedUser> createOfflineUser(UUID uuid) {
        var userFile = getUserFile(uuid);
        var user = new OfflineUser(userFile);

        if (user.getName().isEmpty()) return Optional.empty();

        var cachedUser = new CachedUser(user);
        _userUuidMap.put(uuid, cachedUser);
        _userMap.put(user.getName().get(), cachedUser);

        return Optional.of(cachedUser);
    }

    /**
     * Creates and caches an online user for the given UUID.
     *
     * @param uuid The UUID of the user to create
     * @return The cached user
     */
    private CachedUser createOnlineUser(UUID uuid) {
        var userFile = getUserFile(uuid);
        var user = new User(userFile);

        var cachedUser = new CachedUser(user);
        _userUuidMap.put(uuid, cachedUser);
        _userMap.put(user.getName().get(), cachedUser);

        return cachedUser;
    }

    /**
     * Gets a copy of all currently cached users.
     *
     * @return A set containing all cached users
     */
    public Set<CachedUser> getCachedUsers() {
        return Set.copyOf(_userMap.values());
    }

    /**
     * Removes stale users from the cache to prevent memory leaks.
     * This method is called periodically by a scheduled task.
     */
    public void cleanStaleUsers() {
        var staleUsers = _userMap.values().stream().filter(CachedUser::isStale).collect(Collectors.toSet());

        _userMap.entrySet().removeIf(entry -> staleUsers.contains(entry.getValue()));
        _userUuidMap.entrySet().removeIf(entry -> staleUsers.contains(entry.getValue()));
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/userdata/UserManager.kt
package me.testaccount666.serversystem.userdata

import me.testaccount666.paperktx.extensions.TimeExtensions.toTicks
import me.testaccount666.paperktx.scheduler.skedule.okkero.schedule
import me.testaccount666.serversystem.ServerSystem
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.nio.file.Path
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration.Companion.minutes

/**
 * Manages ServerSystem's user data.
 * This class is responsible for creating, caching, and retrieving user objects,
 * as well as cleaning up stale user data to prevent memory leaks.
 */
class UserManager {
    private val _userMap = ConcurrentHashMap<String, CachedUser>()
    private val _userUuidMap = ConcurrentHashMap<UUID, CachedUser>()

    init {
        ServerSystem.instance.schedule {
            loop(15.minutes.toTicks(), 15.minutes.toTicks()) { cleanStaleUsers() }
        }

        Bukkit.getOnlinePlayers().forEach(::getUserOrNull)
    }

    /**
     * Gets a cached user by player instance.
     *
     * @param player The player to get the user for
     * @return An Optional containing the cached user or empty if the user doesn't exist
     */
    fun getUserOrNull(player: Player): CachedUser? {
        if (player.hasMetadata("NPC")) return CachedUser(_NPC_USER)

        return getUserOrNull(player.uniqueId, true)
    }

    /**
     * Gets a cached user by UUID.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param uuid The UUID of the user to get
     * @return The cached user or null if the user doesn't exist
     */
    fun getUserOrNull(uuid: UUID) = getUserOrNull(uuid, false)

    /**
     * Gets a cached user by UUID.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param uuid            The UUID of the user to get
     * @param forceOnlineUser If the method should only return online users
     * @return The cached user or null if the user doesn't exist
     */
    fun getUserOrNull(uuid: UUID, forceOnlineUser: Boolean): CachedUser? {
        if (!forceOnlineUser && uuid == ConsoleUser.CONSOLE_UUID) return CachedUser(consoleUser)

        _userUuidMap[uuid]?.let {
            if (it.isOfflineUser && forceOnlineUser) return null
            it.updateLastAccessTime()
            return it
        }

        if (Bukkit.getPlayer(uuid) != null) return createOnlineUser(uuid)
        return if (!forceOnlineUser) createOfflineUserOrNull(uuid) else null
    }

    /**
     * Gets a cached user by name.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param name The name of the user to get
     * @return The cached user or null if the user doesn't exist
     */
    fun getUserOrNull(name: String) = getUserOrNull(name, false)

    /**
     * Gets a cached user by name.
     * If the user is already cached, returns the cached instance.
     * If the user is online but not cached, creates and caches a new online user.
     * If the user is offline, creates and caches a new offline user.
     *
     * @param name            The name of the user to get
     * @param forceOnlineUser If the method should only return online users
     * @return The cached user or null if the user doesn't exist
     */
    fun getUserOrNull(name: String, forceOnlineUser: Boolean): CachedUser? {
        _userMap[name]?.let {
            if (it.isOfflineUser && forceOnlineUser) return null
            it.updateLastAccessTime()
            return it
        }

        Bukkit.getPlayer(name)?.let { return createOnlineUser(it.uniqueId) }
        if (forceOnlineUser) return null

        val offlinePlayer = Bukkit.getOfflinePlayer(name).takeIf { it.name != null } ?: return null
        return createOfflineUserOrNull(offlinePlayer.uniqueId)
    }

    /**
     * Creates and caches an offline user for the given UUID.
     *
     * @param uuid The UUID of the user to create
     * @return The cached user or null if the user doesn't exist
     */
    private fun createOfflineUserOrNull(uuid: UUID): CachedUser? {
        val user = OfflineUser(getUserFile(uuid))
        val name = user.getNameOrNull() ?: return null

        return CachedUser(user).also {
            _userUuidMap[uuid] = it
            _userMap[name] = it
        }
    }

    /**
     * Creates and caches an online user for the given UUID.
     *
     * @param uuid The UUID of the user to create
     * @return The cached user
     */
    private fun createOnlineUser(uuid: UUID): CachedUser {
        val user = User(getUserFile(uuid))

        return CachedUser(user).also {
            _userUuidMap[uuid] = it
            _userMap[user.getNameOrNull()!!] = it
        }
    }

    val cachedUsers
        get() = _userMap.values.toSet()

    /**
     * Removes stale users from the cache to prevent memory leaks.
     * This method is called periodically by a scheduled task.
     */
    fun cleanStaleUsers() {
        _userMap.values.removeIf { it.isStale }
        _userUuidMap.values.removeIf { it.isStale }
    }

    companion object {
        @JvmField
        val USER_DATA_PATH: Path = ServerSystem.instance.dataFolder.toPath().resolve("UserData")

        /**
         * Gets the console user instance.
         *
         * @return The console user instance
         */
        val consoleUser = ConsoleUser()
        private val _NPC_USER = NpcUser()
        private fun getUserFile(uuid: UUID) = USER_DATA_PATH.resolve("${uuid}.yml.gz").toFile()
    }
}